engine/CSVParser.cc

```diff
@@ -447,26 +448,62 @@ namespace minsky
                   }
               }
           }
-  
-        v.setXVector(xVector);
-        if (!cminsky().checkMemAllocation(v.numElements()*sizeof(double)))
-          throw runtime_error("memory threshold exeeded");
-        // stash the data into vv tensorInit field
-        v.tensorInit.data.clear();
-        v.tensorInit.data.resize(v.numElements(), spec.missingValue);
-        auto dims=v.tensorInit.dims=v.dims();    
-        for (auto& i: tmpData)
-          {
-            size_t idx=0;
-            assert (dims.size()==i.first.size());
-            assert(dimLabels.size()==dims.size());
-            for (int j=dims.size()-1; j>=0; --j)
-              {
-                assert(dimLabels[j].count(i.first[j]));
-                idx = (idx*dims[j]) + dimLabels[j][i.first[j]];
-              }
-            v.tensorInit.data[idx]=i.second;
-          }
+                  
+        size_t numHyperCubeElems=1;
+        for (auto& i : xVector) numHyperCubeElems*=i.size();
+                           
+        double sparsityRatio =  static_cast<double>(1.0-static_cast<double>(tmpData.size())/numHyperCubeElems); 
+		
+		// for feature 47
+		if (sparsityRatio <= 0.5) 
+	    {
+		  v.index.clear();	
+          v.setXVector(xVector);          
+          if (!cminsky().checkMemAllocation(v.numDenseElements()*sizeof(double)))
+            throw runtime_error("memory threshold exceeded");            
+          // stash the data into vv tensorInit field
+          v.tensorInit.data.clear();
+          v.tensorInit.data.resize(v.numDenseElements(), spec.missingValue);  
+          auto dims=v.tensorInit.dims=v.dims();    
+          for (auto& i: tmpData)
+            {
+              size_t idx=0;
+              assert (dims.size()==i.first.size());
+              assert(dimLabels.size()==dims.size());
+              for (int j=dims.size()-1; j>=0; --j)
+                {
+                  assert(dimLabels[j].count(i.first[j]));
+                  idx = (idx*dims[j]) + dimLabels[j][i.first[j]];
+                }
+              v.tensorInit.data[idx]=i.second;  
+            }
+        }    
+        else 
+	    {	
+          v.index.clear();			
+          v.setXVector(xVector);          
+          if (!cminsky().checkMemAllocation(tmpData.size()*sizeof(double)))
+            throw runtime_error("memory threshold exceeded");	  	  		
+          auto dims=v.tensorInit.dims=v.dims();	
+          v.tensorInit.data.clear();   
```

Comment:
Common code like this should be moved outside the conditional

Reply:
Done. Unfortunately I cannot declare a local index vector conditionally here, so it is declared in the common code whether the data is sparse or not (to be committed shortly).